### PR TITLE
builder/amazon: Add `ebs-init` builder

### PR DESCRIPTION
**Pull request by jen20**
This commit adds a builder that works like EBS builders, except does not create an AMI, and instead is intended to create EBS volumes in an initialized state. For example, the following template can be used to create and export a set of 3 EBS Volumes in a ZFS zpool named `data` for importing by instances running production systems:

```json
{
	"variables": {
		"aws_access_key_id": "{{ env `AWS_ACCESS_KEY_ID` }}",
		"aws_secret_access_key": "{{ env `AWS_SECRET_ACCESS_KEY` }}",
		"region": "{{ env `AWS_REGION` }}",
		"source_ami": "{{ env `PACKER_SOURCE_AMI` }}",
		"vpc_id": "{{ env `PACKER_VPC_ID` }}",
		"subnet_id": "{{ env `PACKER_SUBNET_ID` }}"
	},
	"builders": [{
		"type": "amazon-ebs-volume",
		"access_key": "{{ user `aws_access_key_id` }}",
		"secret_key": "{{ user `aws_secret_access_key` }}",
		"region": "{{user `region`}}",
		"spot_price_auto_product": "Linux/UNIX (Amazon VPC)",

		"ssh_pty": true,
		"instance_type": "t2.medium",
		"vpc_id": "{{user `vpc_id` }}",
		"subnet_id": "{{user `subnet_id` }}",
		"associate_public_ip_address": true,
		"source_ami": "{{user `source_ami` }}",
		"ssh_username": "ubuntu",
		"ssh_timeout": "5m",

		"ebs_volumes": [
			{
				"device_name": "/dev/xvdf",
				"delete_on_termination": false,
				"volume_size": 10,
				"volume_type": "gp2",
				"tags": {
					"Name": "TeamCity-Data1",
					"zpool": "data",
					"Component": "TeamCity"
				}
			},
			{
				"device_name": "/dev/xvdg",
				"delete_on_termination": false,
				"volume_size": 10,
				"volume_type": "gp2",
				"tags": {
					"Name": "TeamCity-Data2",
					"zpool": "data",
					"Component": "TeamCity"
				}
			},
			{
				"device_name": "/dev/xvdh",
				"delete_on_termination": false,
				"volume_size": 10,
				"volume_type": "gp2",
				"tags": {
					"Name": "TeamCity-Data3",
					"zpool": "data",
					"Component": "TeamCity"
				}
			}
		]
	}],
	"provisioners": [
	{
		"type": "shell",
		"start_retry_timeout": "10m",
		"inline": [
			"DEBIAN_FRONTEND=noninteractive sudo apt-get update",
			"DEBIAN_FRONTEND=noninteractive sudo apt-get install -y zfs",
			"lsblk",
			"sudo parted /dev/xvdf --script mklabel GPT",
			"sudo parted /dev/xvdg --script mklabel GPT",
			"sudo parted /dev/xvdh --script mklabel GPT",
			"sudo zpool create -m none data raidz xvdf xvdg xvdh",
			"sudo zpool status",
			"sudo zpool export data",
			"sudo zpool status"
		]
	}
	]
}
```

TODO:

- [x] - Builder implementation
- [ ] - Determine unit testing to be done
- [ ] - Determine acceptance testing to be done
- [x] - Write documentation

### CODE REVIEW

**Comment by mwhooker** [comment]
if any of the steps are unmodified copies of amazon/ebs steps, we can probably move them to /common

**Comment by mwhooker** [comment]
I think this looks pretty good so far. I don't think I actually like the name...

My top proposal is `amazon-ebs-volume` even though it's kinda wordy

**Comment by jen20** [comment]
@mwhooker Any advice as to what type of testing you'd want to see around this builder? I will check for duplicated code.

**Comment by jen20** [comment]
I've rebased this onto master and eliminated the duplicated code.

**Comment by mwhooker** [comment]
cool! will review tomorrow

File: website/source/docs/builders/amazon-ebs-volume.html.md [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+
+### Required:
+
+-   `access_key` (string) - The access key used to communicate with AWS. [Learn
+    how to set this.](/docs/builders/amazon.html#specifying-amazon-credentials)
+
+-   `ami_name` (string) - The name of the resulting AMI that will appear when
+    managing AMIs in the AWS console or via APIs. This must be unique. To help
+    make this unique, use a function like `timestamp` (see [configuration
+    templates](/docs/templates/configuration-templates.html) for more info)
```
Comment (rickard-von-essen) [inline website/source/docs/builders/amazon-ebs-volume.html.md]: Should be deleted.

File: website/source/docs/builders/amazon-ebs-volume.html.md [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+-   `availability_zone` (string) - Destination availability zone to launch
+    instance in. Leave this empty to allow Amazon to auto-assign.
+
+-   `ebs_optimized` (boolean) - Mark instance as [EBS
+    Optimized](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/EBSOptimized.html).
+    Default `false`.
+
+-   `enhanced_networking` (boolean) - Enable enhanced
+    networking (SriovNetSupport) on HVM-compatible AMIs. If true, add
+    `ec2:ModifyInstanceAttribute` to your AWS IAM policy.
```
Comment (rickard-von-essen) [inline website/source/docs/builders/amazon-ebs-volume.html.md]: Drop?

File: website/source/docs/builders/amazon-ebs-volume.html.md [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+    to be set.
+
+-   `windows_password_timeout` (string) - The timeout for waiting for a Windows
+    password for Windows instances. Defaults to 20 minutes. Example value: "10m"
+
+-   `shutdown_behaviour` (string) - Automatically terminate instances on shutdown
+    incase packer exits ungracefully. Possible values are "stop" and "terminate",
+    default is stop.
+
+## Basic Example
```
Comment (rickard-von-essen) [inline website/source/docs/builders/amazon-ebs-volume.html.md]: I think this is a way too big example to serve as a "basic example". I think it should be something like:

```json
{
   "type" : "amazon-ebsinit",
   "secret_key" : "YOUR SECRET KEY HERE",
   "access_key" : "YOUR KEY HERE",
   "region" : "us-east-1",
   "ssh_username" : "ubuntu",
   "instance_type" : "t2.medium",
   "source_ami" : "ami-40d28157",
   "ebs_volumes" : [
      {
         "volume_type" : "gp2",
         "device_name" : "/dev/xvdf",
         "delete_on_termination" : false,
         "tags" : {
            "zpool" : "data",
            "Name" : "Data1"
         },
         "volume_size" : 10
      },
      {
         "volume_type" : "gp2",
         "device_name" : "/dev/xvdg",
         "tags" : {
            "zpool" : "data",
            "Name" : "Data2"
         },
         "delete_on_termination" : false,
         "volume_size" : 10
      },
      {
         "volume_size" : 10,
         "tags" : {
            "Name" : "Data3",
            "zpool" : "data"
         },
         "delete_on_termination" : false,
         "device_name" : "/dev/xvdh",
         "volume_type" : "gp2"
      }
   ]
}
```

File: builder/amazon/ebsvolume/builder.go [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+	return nil, nil
+}
+
+func (b *Builder) Run(ui packer.Ui, hook packer.Hook, cache packer.Cache) (packer.Artifact, error) {
+	config, err := b.config.Config()
+	if err != nil {
+		return nil, err
+	}
+
+	session := session.New(config)
```
Comment (mwhooker) [inline builder/amazon/ebsvolume/builder.go]: probably best to change this and then do all the rest in a separate PR

Comment (mwhooker) [inline builder/amazon/ebsvolume/builder.go]: created issue here for this https://github.com/mitchellh/packer/issues/4093

File: builder/amazon/ebsvolume/builder.go [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+	ec2conn := ec2.New(session)
+
+	// If the subnet is specified but not the AZ, try to determine the AZ automatically
+	if b.config.SubnetId != "" && b.config.AvailabilityZone == "" {
```
Comment (mwhooker) [inline builder/amazon/ebsvolume/builder.go]: it's a shame this code is duplicated in builder/amazon/ebs/builder.go and builder/amazon/instance/builder.go already.

won't consider it in scope of this PR, but if you did want to find a way to reuse this...

**Comment by mwhooker** [comment]
besides @rickard-von-essen's comments, which I agree with, I think this looks pretty good

File: website/source/docs/builders/amazon-ebs-volume.html.md (hunk 2 above)
Comment (jen20) [inline website/source/docs/builders/amazon-ebs-volume.html.md]: You can use enhanced networking on the source instance - although it's uncommon I don't think there is a reason to drop this.

**Comment by jen20** [comment]
OK, all of those should be addressed.

**Comment by jen20** [comment]
I've squashed into one commit and rebased onto master to pick up some changes to SSH.

**Comment by rickard-von-essen** [comment]
@mwhooker I fixed the name in the changelog in ac04315

**Comment by rickard-von-essen** [comment]
A bit late on the ball here, but shouldn't this builder also return a artefact? So you for example could use the `manifest` post-processor to write down the `volume-ids` into a manifest?

**Comment by mwhooker** [comment]
Yeah, definitely. Will create a ticket to get it added


On Fri, Nov 4, 2016 at 4:06 AM Rickard von Essen <notifications@github.com>
wrote:

A bit late on the ball here, but shouldn't this builder also return a
artefact? So you for example could use the manifest post-processor to write
down the volume-ids into a manifest?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/mitchellh/packer/pull/4088#issuecomment-258404317>, or mute
the thread
